### ISSUE TITLE
access services in steps does not work as documented

### ISSUE THREAD

**Issue by an-ky** [buildbot/buildbot#5558]
I implemented a Service for buildbot and tried to access it in my step like it is written in the documentation using `self.master.namedServices['JiraService']`, I get a KeyError when executing the step.

I windmilled around, trial and errored some stuff until I discovered that my service is not listed in the master.namedServices dict

```
2020-10-01 22:46:58+0000 [-] Available Services: metrics,caches,pbmanager,pbListener,WorkerManager,ChangeManager,botmaster,MachineManager,SchedulerManager,user_manager_manager,db,wamp,mq,data,www,debug_services,secrets,services,HTTPClientService_ba45319207e79423b0d9a70670480b9c08d2838d,HTTPClientService_d266dc0717a0e8bb6007fc64b6af7b6c1c723626,HTTPClientService_9f3d49eca4209180517b40215e008f9302995b35
```

Further investigations using manhole unearthed that my service is actually listed in the services object.

```
>>> show(master.namedServices['services'])
data attributes of <buildbot.util.service.BuildbotServiceManager object at 0x7f8330be9358>
                      config_attr : services
                           master : <class 'buildbot.master.BuildMaster'>
                             name : services
                    namedServices : {'MattermostService': <service.mattermost.MattermostService object at 0x7f832eb67da0>, 'BitbucketServerStatusPush': <buildbot.reporters.bitbucketserver.BitbucketServerStatusPush object at 0x7f832eb67470>, 'JiraService': <service.jira_service.JiraService object at 0x7f832eb67d30>, 'MattermostStatusPush': <reporters.mattermost.MattermostStatusPush object at 0x7f832eb52f60>} (4 elements)
                           parent : <class 'buildbot.master.BuildMaster'>
                reconfig_priority : 1000
                          running : 1
                         services : [<service.mattermost.MattermostService object at 0x7f832eb67da0>, <buildbot.reporters.bitbucketserver.BitbucketServerStatusPush object at 0x7f832eb67470>, <service.jira_service.JiraService object at 0x7f832eb67d30>, <reporters.mattermost.MattermostStatusPush object at 0x7f832eb52f60>] (4 elements)
<buildbot.util.service.BuildbotServiceManager object at 0x7f8330be9358>
```` 
(See here the JiraService)

Using `self.master.namedServices["services"].namedServices["JiraService"]` in the step now works, but is this really the intended behavior (aka is the documentation wrong) or did I miss some essential setup/registration step for my service so it gets accessible through `master.namedServices` directly?

**Comment by tardyp** [comment]
namedServices["JiraService"] is an internal API. I don't think you are supposed to use it.

For a JiraService, I would advice to use the SharedService, which is created for that.
You can probably subclass from HttpClientService in order to achieve best reuse.

example here:
https://github.com/buildbot/buildbot/blob/bbe109ff96475b0d45251041ccc65370f30abf72/master/buildbot/util/kubeclientservice.py#L209

If you want to share your code, I will be able to help more.

**Comment by an-ky** [comment]
Hi tardyp,

I'd be glad if you would help me with my code. Here is the current version of my JiraService, which might be rather clumsy.
I'm still fighting a bit with the rendering of secret parameters, too. The parameters user and password are not automatically rendered despite being stated in the `secrets` class variable.

```
from buildbot import config
from buildbot.util.service import BuildbotService
from jira import JIRA
from twisted.python import log
from buildbot.status.builder import SUCCESS


class JiraService(BuildbotService):
    """
    provides access to the jira rest api v2 in a wrapped form

    currently it only supports to get a version_summary which returns
    all issues of a given project and a given fixVersion
    """

    name = "JiraService"

    secrets = [
        "user",
        "passwd"
    ]

    def checkConfig(self, server=None, user=None, passwd=None):
        if server and not isinstance(server, str):
            config.error(
                "server must be a string while it is {}".format(type(server))
            )
        if not (user and passwd):
            config.error(
                "user and passwd are mandatory for JiraService"
            )

        log.msg(f"{self.name} checked")


    @defer.inlineCallbacks
    def reconfigService(self, server=None, user=None, passwd=None, **kwarg):
        user, password = yield self.renderSecrets(user, passwd)
        yield super().reconfigService(**kwarg)

        if server.endswith("/"):
            server = server[:-1]

        self._jira = JIRA(server, auth=(user, password))

        log.msg("Jira service configured")


    @defer.inlineCallbacks
    def version_summary(self, project, fix_version):
        """
        pass jql and fields to jira and return result

        generates a list of dictionaries for each issue
        key : string
            The jira issue key
        self : string
            Link the issue
        summary : string
            The issue's summary
        status : str
            The issue"s status
        """
        issues = yield self._jira.search_issues(f"project={project} and fixVersion={fix_version}")
        return [dict(key=issue.key, self=issue.self, summary=issue.fields.summary, resolution=issue.fields.resolution,
                     status=issue.fields.status)
                for issue in issues]
```

This is the step I use the service

```
from buildbot.plugins import *
from twisted.internet import defer
from buildbot.process.results import SUCCESS
from service.jira_service import JiraService
from service.mattermost import MattermostService
from pytablewriter import MarkdownTableWriter
from twisted.python import log


@util.renderer
def version_without_info(props):
    return props.getProperty("semver").split("+")[0]


class VersionReport(steps.BuildStep):
    renderables = [
        "project",
        "fix_version",
        "channel_id",
    ]

    def __init__(self, channel_id, **kwargs):
        super().__init__(**kwargs)

        self.project = util.Property("jira_project_key")
        self.fix_version = version_without_info
        self.channel_id = channel_id

    @defer.inlineCallbacks
    def run(self):
        """execute the step

        uses the jira service to query all issues for the current project
        and the current version which is determined using prop:semver with
        everything beyond + stripped away.
        The versioning scheme in jira must match this scheme so the service
        is able to query for the proper version.

        Build a table from the received issues and post a message with a short
        text informing about the new release and the table with the issues in
        the RHS-Cards view.
        """

        log.msg("Available Services: " + ','.join([k for k in self.master.namedServices]))
        jira = self.master.namedServices["services"].namedServices["JiraService"]
        mattermost = self.master.namedServices["services"].namedServices["MattermostService"]

        issues_in_version = yield jira.version_summary(self.project,
                                                       self.fix_version)

        log.msg(issues_in_version)
        writer = MarkdownTableWriter()
        writer.table_name = "Resolved Issues"
        writer.headers = ["Issue Key", "Summary", "Status", "Resolution"]
        writer.value_matrix = [[f"[{issue['key']}]({issue['self']})", issue['summary'], issue['status'], issue['resolution']] for issue in
                               issues_in_version]

        yield mattermost.postMessage(
            self.channel_id,
            f"@channel\n# Release {self.fix_version} ist jetzt bereit.\n"
            f"Es wurden insgesamt {len(issues_in_version)} Aufgaben erledigt.\n"
            f"Über den Info-Button gelangt ihr zu weiteren Details.",
            writer.dumps()
        )

        return SUCCESS
```

The part of the documentation which tells to use `master.namedServices` is in the developer section [http://docs.buildbot.net/current/developer/utils.html#buildbot.util.service.BuildbotService](BuildbotService) a but futher down after the secrets passage were the example follows:

_Such services are singletons accessible in nearly every objects of buildbot (buildsteps, status, changesources, etc) using self.master.namedServices[‘<nameOfYourService’]._

```
class MyShellCommand(ShellCommand):

    def getResultSummary(self):
        # access the service attribute
        service = self.master.namedServices['myService']
        return dict(step=u"arg value: %d" % (service.arg1,))

class MyService(BuildbotService):
    name = "myService"

    def checkConfig(self, arg1):
        if not isinstance(arg1, int):
            config.error("arg1 must be an integer while it is %r" % (arg1,))
            return
        if arg1 < 0:
            config.error("arg1 must be positive while it is %d" % (arg1,))

    def reconfigService(self, arg1):
        self.arg1 = arg1
        return defer.succeed(None)

c['schedulers'] = [
    ForceScheduler(
        name="force",
        builderNames=["testy"])]

f = BuildFactory()
f.addStep(MyShellCommand(command='echo hei'))
c['builders'] = [
    BuilderConfig(name="testy",
                  workernames=["local1"],
                  factory=f)]

c['services'] = [
    MyService(arg1=1)
]
```

When using `SharedService` as base, is the object sill stored in `c[services]`?

**Comment by tardyp** [comment]
I didn't remember we had c['services'].
You can directly access them in master.services['JiraService']

I am not sure why the secrets are not rendered. this should work. you may want to do some debug prints or pydb here:
https://github.com/buildbot/buildbot/blob/bbe109ff96475b0d45251041ccc65370f30abf72/master/buildbot/util/service.py#L209

```
        issues_in_version = yield jira.version_summary(self.project,
                                                       self.fix_version)
```
This is a blocking call. This should not be used in main thread, as this will slow down the entire master.
You should rather use a deferToThread call for this. There are plenty of examples in the code.

**Comment by an-ky** [comment]
hi tardyp

thanks for the hints with the blocking calls when calling the jira API. I refactored to use deferToThread when calling the API. Also I could 'resolve' the issue with the secret rendering. Don't know what the problem was, maybe a typo in the parameter or the secrets classvar.

The services are still a bit odd to me. Using `master.services['JiraService']` gives ` builtins.AttributeError: 'BuildMaster' object has no attribute 'service'` so I looked around in BuildMaster.py and saw `service_manager`. Using `master.service_manager.namedServices['JiraService']` works, but will just call the internal API again. On the other hand, using `namedService` is mentioned as the way to go in https://buildbot.readthedocs.io/en/latest/developer/utils.html#buildbot.util.service.BuildbotService but directly as `self.master.namedServices['ServiceName']` from a step.

Greets
